Seemingly incorrect error thrown in the new validator.

Validating the following example gives the error below. Although the profile looks correct. 
See attached example and base profile. 

There is a reproduction project on our staging environment: https://staging.simplifier.net/bellucciengenir/hc40-orderentry.observation


Overall result: FAILURE
[WARNING] Cannot resolve reference ServiceRequest/IHUB.1-284679-202209051330-10-0.65045 (at Observation.basedOn[0], element Observation(https://eng.it/fhir/StructureDefinition/Observation_Eng).basedOn)
[WARNING] Cannot resolve reference Patient/1 (at Observation.subject[0], element Observation(https://eng.it/fhir/StructureDefinition/Observation_Eng).subject)
[WARNING] Cannot resolve reference Encounter/HC40-ER.b829db56-8a2f-4c6b-9949-38c4de0664e0 (at Observation.encounter[0], element Observation(https://eng.it/fhir/StructureDefinition/Observation_Eng).encounter)
[ERROR] Encountered unknown child elements 'value','unit','system','code' (at Observation.valueQuantity[0], element Observation(https://eng.it/fhir/StructureDefinition/Observation_Eng).value[x])
[WARNING] Cannot resolve reference Specimen/IHUB.28467901 (at Observation.specimen[0], element Observation(https://eng.it/fhir/StructureDefinition/Observation_Eng).specimen)

--- a/src/Firely.Fhir.Validation.Compilation.Shared/SchemaBuilders/TypeReferenceBuilder.cs
+++ b/src/Firely.Fhir.Validation.Compilation.Shared/SchemaBuilders/TypeReferenceBuilder.cs
@@ -68,15 +68,36 @@ namespace Firely.Fhir.Validation.Compilation
 #if STU3
             var hasProfileDetails = def.Type.Any(tr => !string.IsNullOrEmpty(tr.Profile) || !string.IsNullOrEmpty(tr.TargetProfile));
 #else
-            var hasProfileDetails = def.Type.Any(tr => tr.Profile.Any() || tr.TargetProfile.Any());
+
 #endif
-            if ((!nav.HasChildren || hasProfileDetails) && def.Type.Count > 0)
+            if (shouldValidateTypeReference(nav))
             {
                 var typeAssertion = ConvertTypeReferences(def.Type);
                 if (typeAssertion is not null)
                     yield return typeAssertion;
             }
         }
+
+        private static bool shouldValidateTypeReference(ElementDefinitionNavigator nav)
+        {
+            var def = nav.Current;
+            //if there are no types, you can't validate anything
+            if (!def.Type.Any()) return false;
+
+            //if there are multiple types (value[x]), we should always validate against the type references.            
+            if (def.Type.Count > 1) return true;
+
+            //if there are no children, we should validate against the type reference, because that's where we find the child constraints.
+            if (!nav.HasChildren) return true;
+
+            //if there are children, we should only validate against the type reference if there are profile details.
+            var hasProfileDetails = def.Type.Any(tr => tr.Profile.Any() || tr.TargetProfile.Any());
+            if (hasProfileDetails) return true;
+
+            //there are no profile details and no children, so we don't need to validate against the type reference, because the whole type is already defined by the profile, and will be automatically validated by going over the children.
+            return false;
+        }
+
 
         public IAssertion? ConvertTypeReferences(IEnumerable<ElementDefinition.TypeRefComponent> typeRefs)
         {

--- a/src/Firely.Fhir.Validation.Compilation.Shared/SchemaBuilders/TypeReferenceBuilder.cs
+++ b/src/Firely.Fhir.Validation.Compilation.Shared/SchemaBuilders/TypeReferenceBuilder.cs
@@ -80,7 +80,7 @@ namespace Firely.Fhir.Validation.Compilation
             //if there are no types, you can't validate anything
             if (!def.Type.Any()) return false;
 
-            //if there are multiple types (value[x]), we should always validate against the type references.            
+            //if there are multiple types (value[x]), we should always validate against the type references, even if there are children (constraints on extension, id etc.)
             if (def.Type.Count > 1) return true;
 
             //if there are no children, we should validate against the type reference, because that's where we find the child constraints.
@@ -94,7 +94,7 @@ namespace Firely.Fhir.Validation.Compilation
 #endif
             if (hasProfileDetails) return true;
 
-            //there are no profile details and no children, so we don't need to validate against the type reference, because the whole type is already defined by the profile, and will be automatically validated by going over the children.
+            //there are no profile details and there are children, so we don't need to validate against the type reference, because the whole type is already included as children in this profiled element.
             return false;
         }
 

--- a/src/Firely.Fhir.Validation.Compilation.Shared/SchemaBuilders/TypeReferenceBuilder.cs
+++ b/src/Firely.Fhir.Validation.Compilation.Shared/SchemaBuilders/TypeReferenceBuilder.cs
@@ -65,11 +65,7 @@ namespace Firely.Fhir.Validation.Compilation
             if (conversionMode == ElementConversionMode.ContentReference) yield break;
 
             var def = nav.Current;
-#if STU3
-            var hasProfileDetails = def.Type.Any(tr => !string.IsNullOrEmpty(tr.Profile) || !string.IsNullOrEmpty(tr.TargetProfile));
-#else
 
-#endif
             if (shouldValidateTypeReference(nav))
             {
                 var typeAssertion = ConvertTypeReferences(def.Type);
@@ -91,7 +87,11 @@ namespace Firely.Fhir.Validation.Compilation
             if (!nav.HasChildren) return true;
 
             //if there are children, we should only validate against the type reference if there are profile details.
+#if STU3
+            var hasProfileDetails = def.Type.Any(tr => !string.IsNullOrEmpty(tr.Profile) || !string.IsNullOrEmpty(tr.TargetProfile));
+#else
             var hasProfileDetails = def.Type.Any(tr => tr.Profile.Any() || tr.TargetProfile.Any());
+#endif
             if (hasProfileDetails) return true;
 
             //there are no profile details and no children, so we don't need to validate against the type reference, because the whole type is already defined by the profile, and will be automatically validated by going over the children.

--- a/test/Firely.Fhir.Validation.Compilation.Tests.Shared/SchemaBuilders/TypeReferenceBuilderTests.cs
+++ b/test/Firely.Fhir.Validation.Compilation.Tests.Shared/SchemaBuilders/TypeReferenceBuilderTests.cs
@@ -8,6 +8,7 @@
 using FluentAssertions;
 using FluentAssertions.Primitives;
 using Hl7.Fhir.Model;
+using Hl7.Fhir.Specification.Navigation;
 using System.Collections.Generic;
 using System.Linq;
 using Xunit;
@@ -104,6 +105,52 @@ namespace Firely.Fhir.Validation.Compilation.Tests
             var converted = convertTypeReference(rc);
             converted.Should().BeOfType<SchemaReferenceValidator>().Which.SchemaUri.
                 Should().Be(new Canonical("http://hl7.org/fhirpath/System.String"));
+        }
+
+        [Fact]
+        public void ValueXWithChildrenShouldHaveATypeRef()
+        {
+            var sd = new StructureDefinition
+            {
+                Snapshot = new()
+            };
+            sd.Snapshot.Element =
+            [
+                new ElementDefinition
+                {
+                    Path = "Patient",
+                    Type =
+                    [
+                        new () { Code = "Patient" },
+                    ]
+                },
+                new ElementDefinition
+                {
+                    Path = "Patient.deceased[x]",
+                    Type =
+                    [
+                        new () { Code = "boolean" },
+                        new () { Code = "dateTime" }
+                    ]
+                },
+                new ElementDefinition
+                {
+                    Path = "Patient.deceased[x].extension",
+                    Type =
+                    [
+                        new () { Code = "Extension"}
+                    ]
+                }
+            ];
+
+            var nav = new ElementDefinitionNavigator(sd);
+            nav.JumpToFirst("Patient.deceased[x]");
+
+            var schema = new TypeReferenceBuilder(_fixture.ResourceResolver).Build(nav);
+            var sa = schema.First().Should().BeOfType<SliceValidator>().Subject;
+            sa.Slices.Should().HaveCount(2)
+                .And.Contain(s => s.Name == "boolean")
+                .And.Contain(s => s.Name == "dateTime");
         }
 
 

